docs/design-docs/02-custom-root-reconciler-clusterrole.md

```diff
@@ -1,82 +1,165 @@
 access to deploy the expected resources.
 
+Additionally, the `RepoSync` API currently requires for users to manage the lifecycle
+of `RoleBinding` objects themselves. The proposed API here is intended to also
+simplify the management of `RoleBinding`s for `RepoSync`s.
+
 See also [#935].
 
 [principle of least privilege]: https://en.wikipedia.org/wiki/Principle_of_least_privilege
 [#935]: https://github.com/GoogleContainerTools/kpt-config-sync/issues/935
 
 ## Design Overview
 
-By providing the name of a user-defined `ClusterRole` for a `RootSync`, a user can
+### `RootSync`s
+
+By providing the name of a list user-defined `RoleRef`s for a `RootSync`, a user can
 override which role Config Sync binds to. This configuration is exposed as a new field
 on `spec.overrides` for a `RootSync`:
 
 ```yaml
+kind: RootSync
+metadata:
+  name: my-root-sync
+  namespace: config-management-system
 spec:
   overrides:
-    clusterRole:
-      name: my-custom-role
+    roleRefs:
+    - kind: ClusterRole # Creates a ClusterRoleBinding for ClusterRole my-cluster-role
+      name: my-cluster-role
+    - kind: ClusterRole # Creates a RoleBinding in my-tenant-namespace for ClusterRole tenant-cluster-role
+      name: tenant-cluster-role
+      namespace: my-tenant-namespace
+    - kind: Role # Creates a RoleBinding in my-tenant-namespace for Role my-tenant-role
+      name: my-tenant-role
+      namespace: my-tenant-namespace
 ```
 
-This role is defaulted to `cluster-admin` in order to stay backwards compatible.
+For `RootSync` objects, a default `cluster-admin` `ClusterRoleBinding` will be applied
+when `spec.override.roleRefs` is empty or nil, for reverse compatibility.
+
+For convenience, `RootSync` reconcilers will also be bound to a base `ClusterRole`
+which gives the reconciler the permissions for basic functionality. This will
+be comparable to the pre-existing base ClusterRole for `RepoSync` reconcilers,
+which includes permissions such as status writing on `RepoSync` objects. Leaving
+this permission to the user to manage would create unneeded toil.
+
+### `RepoSync`s
 
-For the case where a single `ClusterRole` is not expressive enough to configure the
-permissions a user want, you can instead set `disabled: true` on the override object:
+For `RepoSync` objects, `roleRefs` entries will not need or allow specifying a namespace.
+Instead, the `namespace` will always be derived from the `RepoSync` object itself.
+This follows the assumption that `RepoSync` objects are always scoped to a single
+Namespace.
 
 ```yaml
+kind: RepoSync
+metadata:
+  name: my-repo-sync
+  namespace: example-ns
 spec:
   overrides:
-    clusterRole:
-      disabled: true
+    roleRefs:
+    - kind: ClusterRole # Creates a RoleBinding in example-ns bound to ClusterRole my-cluster-role
+      name: my-cluster-role
+    - kind: Role # Creates a RoleBinding in example-ns bound to Role my-tenant-role
+      name: my-tenant-role
```

Comment:
This gives privilege escalation, i.e. someone who only has access to RepoSync to have access to modify Rolebinding. It is not safe to add this feature in.

Reply:
It only grants the Role to the root reconciler's service account, which is reconciled by the reconciler-manager, so it seems not an issue.

Reply:
This does allow escalation to namespace admin IF you have write access to the RepoSync. That's different than how it has worked in the past. But there's another way to look at this.

Historically, we've considered two personas: Platform Admin, and Tenant. Platform Admin controlled the RootSyncs, and Tenants controlled the RepoSyncs, but there's another way to frame the personas by considering who has control over the Sources of Truth.

In practice, we've seen many customers use a single RootSync to provision namespaces, RBAC, and RepoSyncs. In this scenario, the Platform Admin controls both the RootSyncs and the RepoSyncs. With this setup, Tenants need to make PRs to the Platform Admin's git repo in order to make changes to those RepoSyncs. That way the Platform Admin can configure Git to require them to approve any changes to the RepoSync and their permissions.

This setup has the advantage that the reconciler-manager owns the RoleBindings and can ensure that they are deleted in the right order, without requiring the Platform Admin to add any depends-on to block deletion of the RoleBindings.

But you're right, with this change, if the Tenant is given RBAC to write RepoSyncs, they effectively are given admin permission in that namespace, and that's different than previous behavior where write access to RoleBindings needed to be managed separately.

Reply:
I suppose the question here is whether the tenant is allowed to create/update `RepoSync` objects themselves, or if the `RepoSync` is created on their behalf (for example, by a platform admin). The premise of https://github.com/GoogleContainerTools/kpt-config-sync/issues/935 makes the implicit assumption that tenants do not have write access to the RSync object itself. I realize I may be incorrectly concluding that assumption aligns with the design of RSync tenancy.

Given that each RepoSync reconciler has a unique ServiceAccount; if a tenant has permission to create RepoSyncs but not RoleBindings, the tenant would need to ask a privileged user to create RoleBindings on their behalf any time they create a new RepoSync. Is this the intended tenancy UX?

In other words, is the tenancy design intended to be that tenants have permission to manage their own RepoSyncs? Or is it assumed that a privileged user (e.g. platform admin) will be creating RepoSyncs on their behalf? 

---
Also, does the same concern apply to `RootSync`? Or is that not the case because `RootSync` reconcilers are granted `cluster-admin` by default?

Reply:
It's intentional that Tenant only get RBAC access from the admin and can't grant themselves (or the RepoSyncs they own) more permission then what they already have. This is access model of multi-repo. It's also documented in https://cloud.google.com/anthos-config-management/docs/how-to/multiple-repositories. 

This technically applies to RootSync as well, but then RootSync historically already have cluster-admin access.  

> It only grants the Role to the root reconciler's service account, which is reconciled by the reconciler-manager, so it seems not an issue.

Isn't RepoSync is controlled by namespace reconciler? Whatever access the namespace reconciler has, the owner of the RepoSync can get the same access to modify resources via RepoSync.

Reply:
Actually, server-side apply might allow for a hybrid setup. 

What if the Platform Admin managed RepoSyncs centrally in their SoT, and the source just doesn't include the `branch`/`revision` field? Then Tenants and their CI could be given RBAC to manage the RepoSync, but their changes would be reverted for fields specified in the Platform Admin SoT. That would allow the Tenant to update the `branch`/`revision` whenever they need. 

Still, shared control of the spec of a resource is problematic UX we should try to address.

Reply:
FWIW, our preferred model would have been to have Platform Admins own the definitions of both types of RSync. Tenants can change the synced ref by having the RepoSync reference a movable target that is not mainline if they want more controlled deployments - eg, target an OCI image tag `prod` which is manually updated at the source of truth to trigger a deploy. (We abandoned that because config sync could not handle the scale of our number of RepoSyncs, and are now using other, push-based, means to get tenant config into our clusters...).

The model where tenants create RepoSyncs themselves has never been attractive to us.